README.md

```diff
@@ -22,14 +30,83 @@ dependencies: [
 ]
 ```
 
+## Usage
+
+`BaggageContext` is intended to be used in conjunction with the instrumentation of distributed systems. To make this
+instrumentation work, all parties involved operate on the same `BaggageContext` type. These are the three common
+parties, in no specific order, and guidance on how to use `BaggageContext`:
+
+### Frameworks/Libraries
+
+Developers creating frameworks/libraries (e.g. NIO, gRPC, AsyncHTTPClient, ...) which benefit from being instrumented
+should adopt `BaggageContext` as part of their public API. AsyncHTTPClient for example might accept a context like this:
+
+```swift
+let context = BaggageContext()
+client.get(url: "https://swift.org", context: context)
+```
+
+For more information on where to place this argument and how to name it, take a look at the
+["Context-Passing Guidelines"](#Context-Passing-Guidelines).
+
+Generally speaking, framework developers do not care about the specific contents of the `BaggageContext` but need to
+make sure to feed it through the entire system. At cross-cutting boundaries, e.g. right before sending an HTTP
+request, they'd inject the `BaggageContext` into the HTTP headers, allowing context propagation. On the flip-sight, an
+HTTP server should extract the request headers into a `BaggageContext`. Injecting/extracting is part of the
+`swift-tracing` libraries [and documented in its own repository](https://github.com/slashmo/gsoc-swift-tracing).
+
+### Instrumentation
+
+When implementing instrumentation for cross-cutting tools, `BaggageContext` becomes the way you propagate metadata such
+as trace ids. Because each instrument knows what values might be added to the `BaggageContext` they are the ones
+creating `BaggageContextKey` types dictating the type of value associated with each key added to the context:
+
+```swift
+private enum TraceIDKey: BaggageContextKey {
+  typealias Value = String
+}
+
+var context = BaggageContext()
+context[TraceIDKey.self] = "4bf92f3577b34da6a3ce929d0e0e4736"
+print(context[TraceIDKey.self] ?? "new trace id")
+```
+
+**TODO**: Add paragraph on why keys should be private/internal (#25)
+
+To make accessing values a bit more convenient, we encourage you to add computed properties to `BaggageContextProtocol`:
+
+```swift
+extension BaggageContextProtocol {
+  var traceID: String? {
+    get {
+      return self[TraceIDKey.self]
+    }
+    set {
+      self[TraceIDKey.self] = newValue
+    }
+  }
+}
+```
+
+### Distributed System
```

Comment:
I'd rename to 

```suggestion
### End Users - explicit context passing
```

or something like that; It is crucial to not scare away people from this concept. People may not be really fully aware they're building distributed systems ("but i just call the database" etc). We should absolutely normalize passing around contexts, it is not weird or special, it's just status quo of today's server development ecosystem.

Reply:
Please make this sub-section the _first_ of the three.

We should go in increasing order of complexity:

- end users
- library authors
- instrumentation authors

Reply:
Yeah, that's a good point. I agree that distributed systems might be a bit intimidating. The reordering also makes sense to me as that's probably also the decreasing order of the amounts of developers in these fields.